sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,707 @@
+func (s *StoragePackerV2) storeBucket(ctx context.Context, bucket *LockedBucket, depth int) error {
+	marshaledBucket, err := proto.Marshal(bucket.Bucket)
+	if err != nil {
+		return errwrap.Wrapf("failed to marshal bucket: {{err}}", err)
+	}
+
+	compressedBucket, err := compressutil.Compress(marshaledBucket, &compressutil.CompressionConfig{
+		Type: compressutil.CompressionTypeSnappy,
+	})
+	if err != nil {
+		return errwrap.Wrapf("failed to compress packed bucket: {{err}}", err)
+	}
+
+	// Store the compressed value
+	err = s.BucketStorageView.Put(ctx, &logical.StorageEntry{
+		Key:   bucket.Key,
+		Value: compressedBucket,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), physical.ErrValueTooLarge) {
+			if depth > 0 {
+				return errwrap.Wrapf("recursive sharding detected: {{err}}", err)
+			} else if !s.disableSharding {
+				err = s.shardBucket(ctx, bucket, s.GetCacheKey(bucket.Key))
+			}
+		}
+		if err != nil {
+			return errwrap.Wrapf("failed to persist packed storage entry: {{err}}", err)
+		}
+	}
+
+	return nil
+}
+
+// upsert either inserts a new item into the bucket or updates an existing one
+// if an item with a matching key is already present.
+func (p *partitionedRequests) upsertItems() error {
+	if p == nil {
+		return fmt.Errorf("nil partition")
+	}
+	bucket := p.Bucket
+	if bucket == nil {
+		return fmt.Errorf("nil bucket")
+	}
+
+	if bucket.HasShards {
+		return fmt.Errorf("upserting item into sharded bucket")
+	}
+	if bucket.ItemMap == nil {
+		bucket.ItemMap = make(map[string][]byte)
+	}
+
+	for _, r := range p.Requests {
+		// Nil data checked earlier
+		bucket.ItemMap[r.ID] = r.Value.Data
+	}
+	return nil
+}
+
+// Set the value of multiple items, as an efficient but non-atomic operation.
+// Each affected bucket will be written just once.
+// A multierror will be returned if some of the writes fail.
+func (s *StoragePackerV2) PutItem(ctx context.Context, items ...*Item) error {
+	idsSeen := make(map[string]bool, len(items))
+	for idx, item := range items {
+		if item == nil {
+			return fmt.Errorf("nil item at index %v", idx)
+		}
+		if item.ID == "" {
+			return fmt.Errorf("missing ID in item at index %v", idx)
+		}
+		if item.Message != nil {
+			return fmt.Errorf("'Message' is deprecated; use 'Data' instead")
+		}
+		if item.Data == nil {
+			return fmt.Errorf("missing data in item at index %v", idx)
+		}
+		if _, found := idsSeen[item.ID]; found {
+			return fmt.Errorf("duplicate ID %v in item at index %v", item.ID, idx)
+		}
+		idsSeen[item.ID] = true
+	}
+
+	// Data flow:
+	//     (id, data)+
+	// 1. compute hashes
+	//     (id, key=hash(id), data)+
+	// 2. sort and partition by bucket
+	//     bucket NN => (id, key=NN*, data)+
+	// 3. acquire storage locks in storage lock order
+	// 4. process items in each bucket
+
+	requests := s.keysForItems(items)
+
+	// Identify the buckets and acquire their corresponding write-locks
+	retryRequired := true
+	var partition []*partitionedRequests
+	for retryRequired {
```

Comment:
Are there any potential pathological cases where this could just loop for a _really_ long amount of time? Like if there were an issue underlying `s.lockBuckets`? I looked at that method earlier and I can't see any reason that would happen, but just wanted to sanity check here.

Reply:
Or, I suppose, the other thing that could happen could be that it might loop _quickly_ and take up a lot of processing power. Should there be any short sleeps in here to wait for a transient problem with locking buckets to resolve?

Reply:
These are truly innocent questions - :-) I haven't worked on the sharding pieces yet so am unfamiliar.

Reply:
We loop only when some other thread has sharded a bucket between this thread looking it up and acquiring the bucket's lock.  That means that the *other* thread has made progress (it successfully got past this point); the system isn't deadlocked or livelocked.

We were either descheduled, or blocked on that bucket lock, in that case, so it doesn't seem possible for us to burn a lot of CPU time looping here. And, we expect resharding to be relatively uncommon (no more than once per 300KB written, and usually once per 300KB * (# of leaves) of new data written.  So I think that we don't need to worry about fairness in that the expected number of loops is really really close to 1.

Reply:
That makes sense! Thanks for talking it through!